### PR TITLE
feat(FOCUS-378): Estimate shipping methods

### PR DESCRIPTION
**Pull request by AlexanderDevitsky**
## Description
FOCUS-378
Related PR: https://github.com/AbsoluteWebServices/focus-next-theme/pull/196

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/api-client/src/types/GraphQL.ts [hunk 1]
```diff
@@ -6941,7 +6943,7 @@ export type CustomerAvailablePaymentMethodsQueryFocus = BaseQuery & CustomerAvai
...
+export type CustomerAvailableShippingMethodsQuery = { customerCart: { shipping_addresses: Array<Maybe<{ available_shipping_methods?: Maybe<Array<Maybe<{ available: boolean, carrier_code: string, carrier_title: string, error_message?: Maybe<string>, method_code?: Maybe<string>, method_title?: Maybe<string>, amount: { currency?: Maybe<CurrencyEnum>, value?: Maybe<number> }, price_excl_tax: { currency?: Maybe<CurrencyEnum>, value?: Maybe<number> }, price_incl_tax: { currency?: Maybe<CurrencyEnum>, value?: Maybe<number> }, additional_data?: Maybe<string>, description?: Maybe<string> }>>> }>> } };
 
 export type CustomerAvailableShippingMethodsQueryFocus = BaseQuery & CustomerAvailableShippingMethodsQuery;
```
Comment (dmitry-kozachek) [inline packages/api-client/src/types/GraphQL.ts:6948]: @AlexanderDevitsky Let's not modify base types, I created new type bellow, please modify it instead `CustomerAvailableShippingMethodsQueryFocus`

File: packages/api-client/src/types/GraphQL.ts [hunk 2]
```diff
@@ -6959,7 +6961,7 @@ export type GuestAvailableShippingMethodsQueryVariables = Exact<{
...
+export type GuestAvailableShippingMethodsQuery = { cart?: Maybe<{ shipping_addresses: Array<Maybe<{ available_shipping_methods?: Maybe<Array<Maybe<{ available: boolean, carrier_code: string, carrier_title: string, error_message?: Maybe<string>, method_code?: Maybe<string>, method_title?: Maybe<string>, amount: { currency?: Maybe<CurrencyEnum>, value?: Maybe<number> }, price_excl_tax: { currency?: Maybe<CurrencyEnum>, value?: Maybe<number> }, price_incl_tax: { currency?: Maybe<CurrencyEnum>, value?: Maybe<number> }, additional_data?: Maybe<string>, description?: Maybe<string> }>>> }>> }> };
 
 export type GuestAvailableShippingMethodsQueryFocus = BaseQuery & GuestAvailableShippingMethodsQuery;
```
Comment (dmitry-kozachek) [inline packages/api-client/src/types/GraphQL.ts:6966]: same here, modify `GuestAvailableShippingMethodsQueryFocus` instead

File: packages/api-client/src/types/GraphQL.ts [hunk 3]
```diff
@@ -7539,3 +7541,29 @@ export interface FocusProductInterface {
...
+  firstname?: Scalars['String'];
+  lastname?: Scalars['String'];
+  company?: Scalars['String'];
+  street?: Array<Maybe<Scalars['String']>>;
+  city?: Scalars['String']
+  region?: Maybe<Scalars['String']>;
+  region_id?: Maybe<Scalars['Int']>;
+  postcode?: Maybe<Scalars['String']>;
+  country_code: Scalars['String'];
+  telephone?: Scalars['String'];
```
Comment (dmitry-kozachek) [inline packages/api-client/src/types/GraphQL.ts:7555]: let's have all optional attributes use `Maybe`

File: packages/composables/src/factories/useEstimateShippingMethodsFactory.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+export function useEstimateShippingMethodsFactory<SHIPPING_METHOD, API extends PlatformApi = any>(
+  factoryParams: UseEstimateShippingMethodsFactory<SHIPPING_METHOD, API>,
+) {
+  return function useEstimateShippingMethods(id: string = ''): UseEstimateShippingMethods<SHIPPING_METHOD, API> {
+    // @ts-ignore
+    const estimatedShippingMethods: Ref<SHIPPING_METHOD[]> = sharedRef(null, 'UseEstimateShippingMethods-response');
+    const lowerEstimatedShippingMethod: Ref<SHIPPING_METHOD> = sharedRef(null, 'UseEstimateShippingMethods-lowerEstimatedShippingMethod');
+    const ssrKey = id || 'useEstimateShippingMethods';
```
Comment (dmitry-kozachek) [inline packages/composables/src/factories/useEstimateShippingMethodsFactory.ts:23]: `estimatedShippingMethods` and `lowerEstimatedShippingMethod` aren't using `ssrKey`, let's use `ssrKey` for all composable variables

**Review by dmitry-kozachek — CHANGES_REQUESTED** [review]

**Review by youanden — COMMENTED** [review]
Test

**Pull request merged**